### PR TITLE
Add method to extract non-target metrics from a simulation

### PR DESCRIPTION
**Pull request by dominic-irx**
This PR adds the ability for the user to get AUC- and concentration-based metrics that are not target-based. I try where possible to use the target-related functions and structure.

### CODE REVIEW

File: R/exposure_metrics.R [hunk 1]
```diff
@@ -16,17 +16,22 @@ NULL
...
 #'   calculated using `target_time` argument to `PKPDsim::sim`.
 #' @export
 
-calc_auc_from_sim <- function(sim_output, auc_comp) {
-  aucs <- sim_output$y[sim_output$comp == auc_comp]
-  if (length(aucs) == 1) {
-    aucs # i.e., cumulative
-  } else {
-    diff(aucs)
+calc_auc_from_sim <- function(sim_output, auc_comp, target_time, target_type) {
```
Comment (dominic-irx) [inline R/exposure_metrics.R:26]: The original version of these two functions will not work if target type if AUC24 or AUC12 and there are multiple times requested.

For example:

```
target_time <- c(200, 300)
target_type <- "auc24"
```

Will result in a sim_output that looks like:

```
sim_output <- data.frame(
  t = c(174, 200, 274, 300),
  y = c(1, 2, 3, 4),
  comp = 15 # assume AUC comp
)
```

which would then return a list of 3 AUCs using the original function.

File: R/run_trial.R [hunk 2]
```diff
@@ -149,6 +150,49 @@ run_trial <- function(
       )
     }
 
+    # post-processing to get evaluation metrics
+    if(design$evaluation$type %in% target_types_auc) {
```
Comment (jasmineirx) [inline R/run_trial.R:154]:
```suggestion
    if(isTRUE(design$evaluation$type %in% target_types_auc)) {
```
I think this would be enough to allow for a NULL evaluate design, which would simply return the default return objects, and therefore the user wouldn't have to specify an additional argument?

File: R/run_trial.R [hunk 3]
```diff
@@ -149,6 +150,49 @@ run_trial <- function(
...
+      )
+      eval_exposure <- rbind(
+        eval_exposure,
+        data.frame(
+          id = i,
+          time = eval_time,
+          auc = auc_eval
+        )
+      )
+    } else if (design$evaluation$type %in% target_types_conc){
```
Comment (jasmineirx) [inline R/run_trial.R:174]: if we want to allow for both trough and auc outputs, we could change this to an `if` instead of an `else if`

File: R/create_target_design.R [hunk 4]
```diff
@@ -177,3 +177,60 @@ target_types_conc <- c("peak", "cmax", "cmax_1hr", "trough", "cmin", "conc")
...
+#' @export
+create_eval_design <- function(
+    evaltype = mipd_target_types(),
+    time = NULL,
+    when = NULL,
+    offset = NULL,
+    at = NULL,
+    anchor = c("dose", "day")
+) {
+  evaltype <- match.arg(tolower(evaltype), mipd_target_types())
```
Comment (jasmineirx) [inline R/create_target_design.R:198]: do we want to allow multiple types of data to be reported within one simulation? might be a nice flexibility, with minimal code required to support it?

**Review by jasmineirx — COMMENTED** [review]
overall the approach looks good to me. some comments about functionality and argument names, plus some minor code clean up suggestions

File: R/exposure_metrics.R [hunk 5]
```diff
@@ -16,17 +16,23 @@ NULL
...
+  aucs <- sim_output[sim_output$comp == auc_comp, ]
+  if (target_type %in% c("auc24", "auc12")){
+    # want 12 or 24 hour AUC before original times
+    aucs$diff <- c(NA, diff(aucs$y))
+    aucs$diff[aucs$t %in% target_time]
+  } else if (nrow(aucs) == 1 | target_type == "auc_cum"){
+    aucs$y
+  } else { # target is conc or similar
+    diff(aucs$y)
```
[the remark this answers is not in the document]
Comment (dominic-irx) [reply R/exposure_metrics.R:35]: There's a test on line 83 of test-exposure_metrics that passes in a target_type of "conc" but asks to calculate AUC. Not entirely sure why, but this clause guarantees back compatibility

[the remark this answers is not in the document]
Comment (jasmineirx) [reply R/exposure_metrics.R:35]: perhaps thats an error?

**Comment by dominic-irx** [comment]
- Cleaned up minor code issues here and there
- Cleaned up calc_auc_from_sim to be more general and clearer
- Fixed tests that imo did not make sense (can we have multiple targets?)
- Added support for multiple evaluation metric types (i.e. you can now combine auc24 and cmin, see pegasp vignette for an example)

File: R/run_trial.R [hunk 6]
```diff
@@ -151,46 +151,67 @@ run_trial <- function(
...
+            tmp_eval_design$scheme,
+            res$final_regimen
+          )
+          eval_exposure <- rbind(
+            eval_exposure,
+            data.frame(
+              id = i,
+              time = eval_time,
+              value = auc_eval,
+              type = tmp_eval_design$type
```
Comment (jasmineirx) [inline R/run_trial.R:180]:
```suggestion
              type = design_type
```
simpler to keep track of

File: R/run_trial.R [hunk 7]
```diff
@@ -151,46 +151,67 @@ run_trial <- function(
...
-      )
-      eval_exposure <- rbind(
-        eval_exposure,
-        data.frame(
-          id = i,
-          time = eval_time,
-          conc = conc_eval
-        )
-      )
+    if (!is.null(design$evaluation)){
```
Comment (jasmineirx) [inline R/run_trial.R:154]: this is a fairly contained block of code, and the function is already getting quite long. perhaps we can wrap this into a function that returns `eval_exposure`?

it also seems like we might be able to prevent having quite so much near-duplicated code if we do something like this somewhere in here:

```r
if (design_type %in% target_types_auc) {
  f_calc <- calc_auc_from_regimen
else if (design_type %in% target_types_conc) {
  f_calc <- calc_concentration_from_regimen 
}
new_evaluate_exposure <- f_calc(
  regimen = res$final_regimen,
  parameters = pars_true_i, # true patient parameters
  model = design$sim$model,
  target_design = tmp_eval_design,
  covariates = covs
)
```

**Review by jasmineirx — COMMENTED** [review]
a couple suggestions around readability and modularity, let me know what you think!

File: tests/testthat/test-exposure_metrics.R [hunk 8]
```diff
@@ -126,10 +130,10 @@ test_that("handles IOV correctly", {
   pars <- pkbusulfanmccune::parameters()
   covs <- list(AGE = 15, WT = 70, HT = 150, SEX = 0, T_CL_EFF = 0)
   target_a <- create_target_design(
-    targettype = "cum_auc", targetvalue = 10, time = c(0, 24)
+    targettype = "cum_auc", targetvalue = 10, time = c(24)
   )
   target_b <- create_target_design(
-    targettype = "cum_auc", targetvalue = 10, time = c(48, 72)
+    targettype = "cum_auc", targetvalue = 10, time = c(72)
```
[the remark this answers is not in the document]
Comment (dominic-irx) [reply tests/testthat/test-exposure_metrics.R:136]: I thought there can only be one target for the simulation. If you're asking for cumulative AUC, I was interpreting it to mean "AUC from time 0 until [insert time]". Are we supporting as a target "AUC between time X and Y"?

File: R/create_target_design.R [hunk 9]
```diff
@@ -195,42 +195,43 @@ create_eval_design <- function(
...
+          when <- "cmax"
+        } else if (type_eval %in% c("auc24")) {
+          offset <- 24
+          when <- "dose"
+        } else if (type_eval %in% c("auc12")) {
+          offset <- 12
+          when <- "dose"
+        } else { # cum AUC
+          when <- "dose"
+        }
```
[the remark this answers is not in the document]
Comment (dominic-irx) [reply R/create_target_design.R:222]: big fan of this. will also change in create_target_design

**Comment by dominic-irx** [comment]
Changes:

- use switch statement in eval and target design functions
- more concise eval metrics in run_trial

Question: should we allow cumulative AUC with two time values? see the comment for more detail

File: R/create_target_design.R [hunk 10]
```diff
@@ -207,19 +202,14 @@ create_eval_design <- function(
...
-          offset <- 24
-          when <- "dose"
-        } else if (type_eval %in% c("auc12")) {
-          offset <- 12
-          when <- "dose"
-        } else { # cum AUC
-          when <- "dose"
-        }
+        switch(
+          targettype,
```
Comment (jasmineirx) [inline R/create_target_design.R:206]:
```suggestion
          type_eval,
```

i think this copy/paste error is causing errors in CI

at some point we might want to put this logic in a function since it is identical in two places

File: R/run_trial.R [hunk 11]
```diff
@@ -152,65 +152,36 @@ run_trial <- function(
...
         }
-      }
-      # concentration metrics
-      if (any(names(design$evaluation) %in% target_types_conc)){
-        conc_evals <- names(design$evaluation)[names(design$evaluation) %in% target_types_conc]
-        for (design_type in conc_evals){
-          tmp_eval_design <- list(
-            type = design_type,
-            scheme = design$evaluation[[design_type]]
+        exposure_metric <- f_calc(
```
Comment (jasmineirx) [inline R/run_trial.R:165]: what happens if `design_type` is neither in `target_types_auc` nor `target_types_conc`? let's make sure that isn't a blocking error, or that we return an informative message if that's the case!

**Review by jasmineirx — CHANGES_REQUESTED** [review]
good clean up!

regarding the cum_auc, my question was less about the design of the function and more a question about why the test changed (i.e., if it was due to a change you made to the code that the test needed to be updated)

1 small fix, and one tiny edge case requiring handling. otherwise lgtm

**Comment by dominic-irx** [comment]
> good clean up!
> 
> regarding the cum_auc, my question was less about the design of the function and more a question about why the test changed (i.e., if it was due to a change you made to the code that the test needed to be updated)
> 
> 1 small fix, and one tiny edge case requiring handling. otherwise lgtm

Oh, gotcha. Yes, my change assumes cumulative AUC is time 0 to time T, and only one cumulative AUC can be the target.

File: R/run_trial.R [hunk 12]
```diff
@@ -149,6 +150,44 @@ run_trial <- function(
...
+        tmp_eval_design <- list(
+          type = design_type,
+          scheme = design$evaluation[[design_type]]
+        )
+        if (design_type %in% target_types_auc){
+          f_calc <- calc_auc_from_regimen
+        } else if (design_type %in% target_types_conc){
+          f_calc <- calc_concentration_from_regimen
+        } else {
+          warning(paste(design_type, "is not supported yet"))
```
Comment (jasmineirx) [inline R/run_trial.R:165]: this is not a very user friendly method since you would generate a warning every patient.

if all methods accepted in `create_eval_design` are accepted in the code here, that's fine we don't have to handling that edge case. however, looking at `mipd_target_types()` it does not seem like you are supporting all methods yet.

perhaps instead the best place is to confirm that every single method in `create_eval_design` is supported throughout the code base, and instead put the warning *there* if the method is not supported since then the user would get the error only once and before they start a whole trial set up.

**Review by jasmineirx — APPROVED** [review]
